types/ckeditor__ckeditor5-engine/src/model/documentselection.d.ts

```diff
@@ -0,0 +1,61 @@
+import { Collection } from "@ckeditor/ckeditor5-utils";
+import { Emitter, EmitterMixinDelegateChain } from "@ckeditor/ckeditor5-utils/src/emittermixin";
+import EventInfo from "@ckeditor/ckeditor5-utils/src/eventinfo";
+import { PriorityString } from "@ckeditor/ckeditor5-utils/src/priorities";
+import DomEventData from "../view/observer/domeventdata";
+import Document from "./document";
+import Element from "./element";
+import Position from "./position";
+import Range from "./range";
+import { Marker } from "./markercollection";
+
+export default class DocumentSelection implements Emitter {
+    readonly anchor: Position | null;
+    readonly focus: Position | null;
+    readonly hasOwnRange: boolean;
+    readonly isBackward: boolean;
+    readonly isCollapsed: boolean;
+    readonly isGravityOverridden: boolean;
+    readonly markers: Collection<Marker>;
+    readonly rangeCount: number;
+
+    constructor(doc: Document);
+    containsEntireContent(element?: Element): boolean;
+    destroy(): void;
+    getAttribute(key: string): string | number | boolean | undefined;
+    getAttributeKeys(): Generator<string>;
+    getAttributes(): Generator<[string, string | number | boolean]>;
```

Comment:
This signature does not match the `tomap` typing:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/495730c9aca0e0bf1b43b90bdfffb35ddc8d14f6/types/ckeditor__ckeditor5-utils/src/tomap.d.ts#L9

As can be seen in [LinkCommand](https://github.com/ckeditor/ckeditor5/blob/d64ce31825d8e0c68eebd18841af740a97ec40a0/packages/ckeditor5-link/src/linkcommand.js#L192), typings must be in a way, that you may call:

```
toMap(selection.getAttributes());
```

The signature of `toMap` itself also seems to be wrong. Its arguments are directly handed over to `new Map()` as the code and the documentation states. Derived from ES2015 typings, which state `new Map()` as:

```
new<K, V>(entries?: readonly (readonly [K, V])[] | null): Map<K, V>;
```

I suggest changing the signature of `toMap` to:

```
export default function toMap<K, V>(data: readonly (readonly [K, V])[] | null): Map<K, V>;
```

and the signature of `getAttributes()` to:

```
getAttributes(): ([string, string])[];
```

Reply:
Hello, @mmichaelis. Thanks for your observations.

About `toMap`, according to docs, it supports `Record<any any>`, Array<[any, any]>, and `Map<any, any>`
https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-utils/src/tomap.js#L16

But, about the `getAttributes` it seems you are right. Unfortunately jsdocs from ckeditor are a bit vague to me; `@returns {Iterable.<*>}` made me think we were talking about a generator. Given this, I'll check all these `Generators` and make sure if they actually are just plain arrays.

Reply:
Hi @fedemp,

Thanks for having a look at it. My experience regarding the JSDoc is similar to yours. Sometimes it is not even vague but wrong. By the way: Thanks for contributing. I really honor your effort. Hope, this PR will soon be merged.

Reply:
@mmichaelis I'm about to force push so it will be harder to note the differences. That's why I'm copying here a diff.
I searched for every occurrence of `Generator` in my definitions, and compared it to the actual code.

```diff
diff --git i/types/ckeditor__ckeditor5-engine/src/model/documentfragment.d.ts w/types/ckeditor__ckeditor5-engine/src/model/documentfragment.d.ts
index 9372204fee..9566b8ea7c 100644
--- i/types/ckeditor__ckeditor5-engine/src/model/documentfragment.d.ts
+++ w/types/ckeditor__ckeditor5-engine/src/model/documentfragment.d.ts
@@ -1,5 +1,7 @@
+import Element from "./element";
 import Node from "./node";
 import Range from "./range";
+import Text from "./text";
 
 export default class DocumentFragment implements Iterable<Node> {
     readonly childCount: number;
@@ -9,11 +11,11 @@ export default class DocumentFragment implements Iterable<Node> {
     readonly parent: null;
     readonly root: DocumentFragment;
 
-    [Symbol.iterator](): Iterator<Node>;
+    [Symbol.iterator](): Iterator<Element|Text>;
     getChild(index: number): Node | null;
     getChildIndex(node: Node): number | null;
     getChildStartOffset(node: Node): number | null;
-    getChildren(): Generator<Node>;
+    getChildren(): IterableIterator<Element|Text>;
     getNodeByPath(relativePath: number[]): Node | DocumentFragment;
     getPath(): [];
     is(type: string): boolean;
diff --git i/types/ckeditor__ckeditor5-engine/src/model/documentselection.d.ts w/types/ckeditor__ckeditor5-engine/src/model/documentselection.d.ts
index f5fe850509..d8a51a1f71 100644
--- i/types/ckeditor__ckeditor5-engine/src/model/documentselection.d.ts
+++ w/types/ckeditor__ckeditor5-engine/src/model/documentselection.d.ts
@@ -8,6 +8,7 @@ import Element from "./element";
 import Position from "./position";
 import Range from "./range";
 import { Marker } from "./markercollection";
+import Selection from "./selection";
 
 export default class DocumentSelection implements Emitter {
     readonly anchor: Position | null;
@@ -23,14 +24,14 @@ export default class DocumentSelection implements Emitter {
     containsEntireContent(element?: Element): boolean;
     destroy(): void;
     getAttribute(key: string): string | number | boolean | undefined;
-    getAttributeKeys(): Generator<string>;
-    getAttributes(): Generator<[string, string | number | boolean]>;
+    getAttributeKeys(): IterableIterator<string>;
+    getAttributes(): IterableIterator<[string, string | number | boolean]>;
     getFirstPosition(): Position | null;
     getFirstRange(): Range | null;
     getLastPosition(): Position | null;
     getLastRange(): Range | null;
     getRanges(): Generator<Range>;
-    getSelectedBlocks(): Generator<Element>;
+    getSelectedBlocks(): ReturnType<Selection['getSelectedBlocks']>;
     getSelectedElement(): Element | null;
     hasAttribute(key: string): boolean;
     is(type: string): boolean;
diff --git i/types/ckeditor__ckeditor5-engine/src/model/element.d.ts w/types/ckeditor__ckeditor5-engine/src/model/element.d.ts
index 66895df16b..1a72cc3977 100644
--- i/types/ckeditor__ckeditor5-engine/src/model/element.d.ts
+++ w/types/ckeditor__ckeditor5-engine/src/model/element.d.ts
@@ -14,7 +14,7 @@ export default class Element extends Node {
     getChild(index: number): Element | Text;
     getChildIndex(node: Node): number;
     getChildStartOffset(node: Node): number;
-    getChildren(): Generator<Element | Text>;
+    getChildren(): IterableIterator<Element | Text>;
     getNodeByPath(relativePath: number[]): Element | Text;
     is(type: string, name?: string): boolean;
     offsetToIndex(offset: number): number;
diff --git i/types/ckeditor__ckeditor5-engine/src/model/node.d.ts w/types/ckeditor__ckeditor5-engine/src/model/node.d.ts
index 7a355c096c..014c913cb4 100644
--- i/types/ckeditor__ckeditor5-engine/src/model/node.d.ts
+++ w/types/ckeditor__ckeditor5-engine/src/model/node.d.ts
@@ -19,8 +19,8 @@ export default class Node {
     constructor(attrs?: Parameters<typeof toMap>[0]);
     getAncestors(options?: { includeSelf?: boolean; parentFirst?: boolean }): Array<Element | DocumentFragment>;
     getAttribute(key: string): string | undefined;
-    getAttributeKeys(): Generator<string>;
-    getAttributes(): Generator<[string, string | number | boolean]>;
+    getAttributeKeys(): IterableIterator<string>;
+    getAttributes(): IterableIterator<[string, string | number | boolean]>;
     getCommonAncestor(node: Node, options?: { includeSelf?: boolean }): Element | DocumentFragment | null;
     getPath(): number[];
     hasAttribute(key: string): boolean;
diff --git i/types/ckeditor__ckeditor5-engine/src/model/selection.d.ts w/types/ckeditor__ckeditor5-engine/src/model/selection.d.ts
index bef4379fbd..bf6d51673b 100644
--- i/types/ckeditor__ckeditor5-engine/src/model/selection.d.ts
+++ w/types/ckeditor__ckeditor5-engine/src/model/selection.d.ts
@@ -25,8 +25,8 @@ export default class Selection implements Emitter {
     );
     containsEntireContent(element: Element): boolean;
     getAttribute(key: string): string | boolean | number | undefined;
-    getAttributeKeys(): Generator<string>;
-    getAttributes(): Generator<[string, string | boolean | number]>;
+    getAttributeKeys(): IterableIterator<string>;
+    getAttributes(): IterableIterator<[string, string | boolean | number]>;
     getFirstPosition(): Position | null;
     getFirstRange(): Range | null;
     getLastPosition(): Position | null;
diff --git i/types/ckeditor__ckeditor5-engine/src/model/textproxy.d.ts w/types/ckeditor__ckeditor5-engine/src/model/textproxy.d.ts
index 168795f227..bcf50d734c 100644
--- i/types/ckeditor__ckeditor5-engine/src/model/textproxy.d.ts
+++ w/types/ckeditor__ckeditor5-engine/src/model/textproxy.d.ts
@@ -1,6 +1,7 @@
 import DocumentFragment from "./documentfragment";
 import Element from "./element";
 import Node from "./node";
+import Text from "./text";
 
 export default class TextProxy {
     readonly data: string;
@@ -15,8 +16,8 @@ export default class TextProxy {
 
     getAncestors(options?: { includeSelf?: boolean; parentFirst?: boolean }): Array<Element | DocumentFragment>;
     getAttribute(key: string): string | number | boolean;
-    getAttributeKeys(): Generator<string>;
-    getAttributes(): Generator<string, string | number | boolean>;
+    getAttributeKeys(): ReturnType<Text['getAttributeKeys']>;
+    getAttributes(): ReturnType<Text['getAttributes']>;
     getPath(): number[];
     hasAttribute(key: string): boolean;
     is(type: string): boolean;
diff --git i/types/ckeditor__ckeditor5-engine/src/view/documentfragment.d.ts w/types/ckeditor__ckeditor5-engine/src/view/documentfragment.d.ts
index d3e0e863ca..3b55d71330 100644
--- i/types/ckeditor__ckeditor5-engine/src/view/documentfragment.d.ts
+++ w/types/ckeditor__ckeditor5-engine/src/view/documentfragment.d.ts
@@ -15,6 +15,6 @@ export default class DocumentFragment implements Iterable<Node> {
     _removeChildren(index: number, howMany?: number): Node[];
     getChild(index: number): Node;
     getChildIndex(node: Node): number;
-    getChildren(): Generator<Node>;
+    getChildren(): IterableIterator<Node>;
     is(type: string): boolean;
 }
diff --git i/types/ckeditor__ckeditor5-engine/src/view/element.d.ts w/types/ckeditor__ckeditor5-engine/src/view/element.d.ts
index ab21da225d..ddf163ee36 100644
--- i/types/ckeditor__ckeditor5-engine/src/view/element.d.ts
+++ w/types/ckeditor__ckeditor5-engine/src/view/element.d.ts
@@ -1,4 +1,5 @@
 import Node from "./node";
+import StylesMap from "./stylesmap";
 
 export default abstract class Element extends Node {
     readonly childCount: number;
@@ -18,18 +19,18 @@ export default abstract class Element extends Node {
     ): Element | null;
     getAttribute(key: string): string | undefined;
     getAttributeKeys(): Generator<string>;
-    getAttributes(): Generator<[string, string]>;
+    getAttributes(): Generator<[string, unknown]>;
     getChild(index: number): Node;
     getChildIndex(node: Node): number;
-    getChildren(): Generator<Node>;
-    getClassNames(): Generator<string>;
-    getCustomProperties(): Generator<[string, string]>;
+    getChildren(): IterableIterator<Node>;
+    getClassNames(): IterableIterator<string>;
+    getCustomProperties(): Generator<[string, any]>;
     getCustomProperty(key: string | symbol): any;
     getFillerOffset(): number | null;
     getIdentity(): string;
     getNormalizedStyle(property: string): Record<string, string> | string | undefined;
     getStyle(property: string): string | undefined;
-    getStyleNames(): Generator<string>;
+    getStyleNames(): ReturnType<StylesMap['getStyleNames']>
     hasAttribute(key: string): boolean;
     hasClass(className: string): boolean;
     hasStyle(property: string): boolean;
```

Reply:
Thanks. It works. I especially like the usage of the `ReturnType` utility type. This may make things easier, once the API needs to be adapted.